[ur] Add experimental feature contribution docs (#546)
oneapi-src/unified-runtime: scripts/core/CONTRIB.rst

Conformance Testing
-------------------

For contributions to the core specification conformance tests *should* be
included as part of your change. The conformance tests can be found
under ``test/conformance/<component>``, where component refers to the API
object an entry-point belongs to i.e. platform, enqueue, device.

The conformance tests *should* ideally include end-to-end testing of all the
changes to the specification if possible. At minimum, they *must* cover at
least one test for each of the possible error codes returned, excluding any
disaster cases like ${X}_RESULT_ERROR_OUT_OF_HOST_MEMORY or similar.

Conformance tests *must* not make assumptions about the adapter under test.
Tests fixtures or cases *must* query for support of optional features and skip
testing if unsupported by the adapter.

Experimental Features
=====================

.. warning::

    Experimental features:

    *   May be replaced, updated, or removed at any time.
    *   Do not require maintaining API/ABI stability of their own additions
        over time.
    *   Do not require conformance testing of their own additions.

Experimental features *must* be defined in two files:

> *   ``scripts/core/EXP-FEATURE.rst`` document describing the experimental

Reviewer: These files do not exist. Is that an omission, or is the first person creating an experimental feature supposed to create them?

Author: The intention is for each experimental feature to create a new pair of files describing only that feature. I'll have a go at refining this section to make that clearer.

Reviewer: Yes. Thanks!

<%
    OneApi=tags['$OneApi']
    x=tags['$x']
    X=x.upper()
%>
====================
 Contribution Guide
====================

This contribution guide explains how to contribute to the Unified Runtime
project and what processes you must follow in order to have your changes
accepted into the project.

.. important::

    Before making a contribution you *should* determine if the change should be
    made directly to the core specification or introduced as an experimental
    feature. The criteria we use to make this distinction are as follows:

    *   The feature exists to enable an experimental feature in a parallel
        language runtime being built on top of Unified Runtime.

    *   The design phase of the feature is expected to span multiple oneAPI
        releases.

    *   A proof of concept implementation exists for a single adapter but

>         multiple adapters are intended to be supported.

Reviewer 2: ```suggestion
        multiple adapters are intended to be supported. If you are unsure whether the feature can be supported by certain adapters then seek the advise of an appropriate stakeholder. It is important to consider as early as possible whether the feature is appropriate for other adapters and seek the input of relevant stakeholders for each adapter in order to ensure a portable implementation.
```

Author: I have added a modified version of this.

    for clarification.

Generating Source
=================

The specification and many other components in the Unified Runtime repository
are generated from a set of YAML_ files which are used as inputs to a Mako_
based templating system. The YAML_ file syntax is defined `YAML syntax`_. To
generate the outputs of the Mako_ templates a build directory must be
configured, instructions are available in the `README
<https://github.com/oneapi-src/unified-runtime/blob/main/README.md>`_ file.
Upon successfully configuring a build directory, generate the outputs with the
following command (or suitable build system equivalent):

.. code-block:: console

    $ cmake --build build --target generate

.. _YAML: https://yaml.org/
.. _Mako: https://www.makotemplates.org/
.. _YAML syntax:
   https://github.com/oneapi-src/unified-runtime/blob/main/scripts/YaML.md

Writing YAML
============

Please read the :ref:`core/INTRO:Naming Convention` section prior to making a
contribuition and refer to the `YAML syntax`_ for specifics of how to define
the required contructs.

When writing ``*.yml`` files and ``ur`` or ``UR`` should exist in the output

> use ``$${'x'}`` or ``$${'X'}`` respectively. These will be replaced while

Reviewer 3: I believe it's not rendered well. I can see on your preview:
```
$X}
```

Author: That's odd, I see this
![image](https://github.com/oneapi-src/unified-runtime/assets/1625763/2b5c5d33-a692-4588-86cc-d3ca7ea0c13f)

Author: In both Firefox and Safari on macOS, which browser are you using @lukaszstolarczuk ?

Author: And in Chrome on Windows ![image](https://github.com/oneapi-src/unified-runtime/assets/1625763/b9b2bd94-9501-40ad-ba9f-edcc02b260f0)

Reviewer 3: hmm.. I'm not sure what happened (something changed?), but now it looks ok - please disregard

*   Pointer to handle arguments, such as out parameters, are prefix with ``ph``
    i.e. ``phQueue``.

Forks and Pull Requests
=======================

To submit a pull request to Unified Runtime, you must first create your own
personal fork of the project and submit your changes to a branch. By convention
we name our branches ``<your_name>/<short_description>``, where the description
indicates the intent of your change. You can then raise a pull request
targeting ``oneapi-src/unified-runtime:main``.

When making changes to the specification you *must* commit all changes to files
in the repository as a result of `Generating Source`_.

Before your pull request is merged it *must* pass all jobs in the GitHub
Actions workflow and *must* be reviewed by no less than two code owners.

.. hint::

    When rebasing a branch on top of ``main`` results in merged conflicts it is
    recommended to resolve conflicts in the ``*.yml`` files then `Generating
    Source`_. This will automatically resolve conflicts in the generated source
    files, leaving only conflicts in non-generated source files to be resolved,
    if any.

Core Features
=============

> Core features are defined in the ``*.yml`` files in the `scripts/core

Reviewer 2: At the moment the clearest definition of a "core feature" I see is that it is not a "experimental feature" at the beginning of this guide. In the "core feature" section a definition is not given but a list of examples are: in the .yml. "not an experimental feature" would be fine if all core features were the same, but it seems atm this isn't really the case since some core features are considered optional: it would be good to define this for clarity. 
I think it would also be useful to clarify for everyone working on it what the main goals of UR are, in order of importance, so we can all take this into account when working on features. After some discussion this seems to be
1) Most important: maximize portability of UR features across adapters within a Parallel Language Runtime (PLR). Maybe this can be thought of as the original "PI" part of UR.
2) Secondary and not yet actively taken into account very much: maximize portability of UR features across different PLRs (e.g. particularly SYCL and OPENMP and more in the future)
Here is an attempt based on my understanding:

```suggestion
Unlike an experimental feature, a core specification feature is considered to have a stable API. Note that a core feature does not have to be implemented in all adapters within a PLR. If a core feature is not implemented in all adapters within a PLR then it is considered an optional core feature. Optional features should be avoided as much as possible to maximize portability across adapters. A core feature also does not have to be implemented within all PLR's (e.g. SYCL or OPENMP or others that may be supported in the future.). Features should also aim to be as portable as possible across PLRs but currently this is a secondary concern with respect to portability across adapters within a PLR.
```

Author: This is a good point and I'll make sure to add the clarification although I'll probably tweak the wording.

Author: I have added a modified version of this.

    please seek the advice of an appropriate stakeholder or ask the Unified
    Runtime team via the `GitHub issue tracker
    <https://github.com/oneapi-src/unified-runtime/issues/new>`_.

Generating Source
=================

The specification and many other components in the Unified Runtime repository
are generated from a set of YAML_ files which are used as inputs to a Mako_
based templating system. The YAML file syntax is defined in `YAML syntax`_. To
generate the outputs of the Mako templates a build directory must be
configured, instructions are available in the `README
<https://github.com/oneapi-src/unified-runtime/blob/main/README.md>`_ file.
Upon successfully configuring a build directory, generate the outputs with the
following command (or suitable build system equivalent):

.. code-block:: console

    $ cmake --build build --target generate

.. _YAML: https://yaml.org/
.. _Mako: https://www.makotemplates.org/
.. _YAML syntax:
   https://github.com/oneapi-src/unified-runtime/blob/main/scripts/YaML.md

Writing YAML
============

Please read the :ref:`core/INTRO:Naming Convention` section prior to making a

> contribuition and refer to the `YAML syntax`_ for specifics of how to define

Reviewer 4: `contribuition` -> `contribution`

> the required contructs.

Reviewer 5: Typo on `contructs`?

`Generating Source`_.

Additionally, the following conventions *must* be followed for function
arguments:

*   Argument names are ``camelCase``.
*   Arguments with pointer types are prefixed with ``p`` for each pointer in
    the type i.e. ``char *pMessage``, ``char **ppMessage``, etc.
*   Handle arguments are prefixed with ``h`` i.e. ``hQueue``.
*   Pointer to handle arguments, such as out parameters, are prefix with ``ph``
    i.e. ``phQueue``.

Forks and Pull Requests
=======================

To submit a pull request to Unified Runtime, you must first create your own
personal fork of the project and submit your changes to a branch. By convention
we name our branches ``<your_name>/<short_description>``, where the description
indicates the intent of your change. You can then raise a pull request
targeting ``oneapi-src/unified-runtime:main``.

When making changes to the specification you *must* commit all changes to files
in the repository as a result of `Generating Source`_.

Before your pull request is merged it *must* pass all jobs in the GitHub
Actions workflow and *must* be reviewed by no less than two code owners.

> .. hint::

Reviewer 2: For me this is not rendering correctly?

Author: I may be worth force force refreshing for browser.

    When rebasing a branch on top of ``main`` results in merged conflicts it is
    recommended to resolve conflicts in the ``*.yml`` files then `Generating
    Source`_. This will automatically resolve conflicts in the generated source
    files, leaving only conflicts in non-generated source files to be resolved,
    if any.

Core Features
=============

A core feature *must* have a stable API/ABI and *should* strive to be supported
across all adapters. However, core features *may* be optional and thus only
supported in one or more adapters. A core feature *should* also strive to
enable similar functionality in parallel language runtime (such as SYCL,
OpenMP, ...) where possible although this is a secondary concern.

> .. hint::

Reviewer 2: same here: not rendering correctly in github render of this rst.

    Optional features should be avoided as much as possible to maximize

>     portability across adapters and reducing the overhead required to use make

Reviewer 2: ```suggestion
    portability across adapters and reduce the overhead required to make
```

*   `scripts/core/enqueue.yml`_ defines commands which can be enqueued on a
    queue object.
*   `scripts/core/runtime.yml`_ defines global symbols pertaining to
    initialization and tear down of the entire runtime.
*   `scripts/core/registry.yml`_ contains an enumeration of all entry-points,
    past and present, for use in the XPTI tracing framework. It is
    automatically updated so shouldn't require manual editing.
*   ``scripts/core/exp-<feature>.yml`` see `Experimental Features`_.

.. _scripts/core/common.yml:
   https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/common.yml
.. _scripts/core/enqueue.yml:
   https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/enqueue.yml
.. _scripts/core/runtime.yml:
   https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/runtime.yml
.. _scripts/core/registry.yml:
   https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/registry.yml

Core Optional Features
----------------------

Optional core features *must* be supported in at least one adapter. Support for
an optional core feature *must* be programmatically exposed to the user via

> boolean query call to ${x}DeviceGetInfo and a new enumerator of the form

Reviewer 2: Should this be
```suggestion
boolean query call to ``${x}DeviceGetInfo`` and a new enumerator of the form
```
?

Author: Unfortunately these are run through some scripts the replace `${x}` with a cross-reference that breaks when wrapped in inline monospace markup otherwise I would have done this.

``UR_DEVICE_INFO_<FEATURE_NAME>_SUPPORT`` in ${x}_device_info_t.

Conformance Testing
-------------------

For contributions to the core specification conformance tests *should* be
included as part of your change. The conformance tests can be found
under ``test/conformance/<component>``, where component refers to the API
object an entry-point belongs to i.e. platform, enqueue, device.

The conformance tests *should* ideally include end-to-end testing of all the
changes to the specification if possible. At minimum, they *must* cover at
least one test for each of the possible error codes returned, excluding any

> disaster cases like ${X}_RESULT_ERROR_OUT_OF_HOST_MEMORY or similar.

Reviewer 2: ```suggestion
disaster cases like ``${X}_RESULT_ERROR_OUT_OF_HOST_MEMORY`` or similar.
```
